File: app/src/main/java/mariuszbaleczny/compass/CompassFragment.java
package mariuszbaleczny.compass;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.location.Location;
import android.os.Bundle;
import android.provider.Settings;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import mariuszbaleczny.compass.custom.CustomEditText;
import mariuszbaleczny.compass.custom.CustomEditTextActionEditor;
import mariuszbaleczny.compass.custom.CustomEditTextWatcher;
import mariuszbaleczny.compass.location.CompassToLocationProvider;
import mariuszbaleczny.compass.location.LocationHelper;

public class CompassFragment extends Fragment implements CompassToLocationProvider.CompassToLocationListener,
        CustomEditTextWatcher.OnCoordinateChangeListener {

    public static final String FRAGMENT_TAG = "CompassFragment";
    private static final String LOCATION_PROVIDER = "LocationProvider";
    private static final int REQUEST_CODE_SETTINGS = 0;

    public static CompassFragment newInstance() {
        return new CompassFragment();
    }

    private LocationHelper customTargetLocation;
    private Compass compass;
    private TextView titleTextView;
    private TextView subtitleTextView;
    private TextInputLayout latitudeTextInputLayout;
    private TextInputLayout longitudeTextInputLayout;
    private CustomEditText latitudeEditText;
    private CustomEditText longitudeEditText;

    private CompassToLocationProvider compassToLocationProvider;

    private View.OnClickListener subtitleOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            setupLayoutOnLocationServicesCheckUp();
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        customTargetLocation = new LocationHelper(new Location(LOCATION_PROVIDER));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_compass, container, false);
        setupLayoutElements(view);
        setupOnCompassSensorPresenceTest();
        setupCoordinatesInputField();
        setCoordinatesEditTextEnabled(false);
        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (Utils.isCompassSensorPresent(getActivity())) {
            setupLayoutOnLocationServicesCheckUp();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (compassToLocationProvider != null) {
            compassToLocationProvider.stopIfStarted();
        }
    }

    private void setupLayoutElements(View v) {
        ImageView compassNeedleView = (ImageView) v.findViewById(R.id.fragment_compass_needle);
        ImageView compassRoseView = (ImageView) v.findViewById(R.id.fragment_compass_rose);
        compass = new Compass(compassRoseView, compassNeedleView);

        titleTextView = (TextView) v.findViewById(R.id.fragment_compass_title_text_view);
        subtitleTextView = (TextView) v.findViewById(R.id.fragment_compass_subtitle_text_view);

        latitudeEditText = (CustomEditText) v.findViewById(R.id.fragment_compass_latitude_edit_text);
        longitudeEditText = (CustomEditText) v.findViewById(R.id.fragment_compass_longitude_edit_text);
        latitudeTextInputLayout = (TextInputLayout) v.findViewById(R.id.fragment_compass_latitude_text_input);
        longitudeTextInputLayout = (TextInputLayout) v.findViewById(R.id.fragment_compass_longitude_text_input);
    }

    private void setupOnCompassSensorPresenceTest() {
        if (!Utils.isCompassSensorPresent(getActivity())) {
            setTitleTextView(getString(R.string.compass_not_detected_title), Color.RED);
            setSubtitleTextView("", null);
            setCoordinatesEditTextEnabled(false);
        } else {
            compassToLocationProvider = new CompassToLocationProvider(getActivity());
            compassToLocationProvider.setCompassToLocationListener(this);
        }
    }

    private void setupCoordinatesInputField() {
        if (compassToLocationProvider != null) {
            CustomEditTextActionEditor latitudeActionEditor = new CustomEditTextActionEditor(
                    getActivity(), longitudeEditText);
            CustomEditTextActionEditor longitudeActionEditor = new CustomEditTextActionEditor(
                    getActivity(), latitudeEditText);

            latitudeEditText.setOnEditorActionListener(latitudeActionEditor);
            longitudeEditText.setOnEditorActionListener(longitudeActionEditor);
            latitudeEditText.addTextChangedListener(new CustomEditTextWatcher(true, this));
            longitudeEditText.addTextChangedListener(new CustomEditTextWatcher(false, this));
        }
    }

    private void setupLayoutOnLocationServicesCheckUp() {
        if (Utils.isLocationServicesEnabled(getActivity())) {
            setTitleTextView(getString(R.string.point_north_title), Color.BLACK);
            compassToLocationProvider.startIfNotStarted();
            setLayoutElementsOnProvider(true);
        } else {
            setTitleTextView("", Color.BLACK);
            compassToLocationProvider.stopIfStarted();
            setLayoutElementsOnProvider(false);
            buildAndShowLocationServicesDialog();
        }
    }

    @Override
    public void onCompassPointerRotate(int roseAngle, int needleAngle) {
        compass.rotateRose(roseAngle);
        compass.rotateNeedle(needleAngle);
    }

    /**
     *  Method handles actions respectively for input coordinate entered in EditText.
     *  When coordinate argument is Double.NaN it means EditText field is empty, so
     *  error will be disabled and target location set to null (reset).
     *  Proper coordinate argument will be set to customTargetLocation and its Location
     *  will be passed to compassToLocationProvider. Passing a null will cause resetting
     *  targetLocation, which is intended.
     *  Sets TitleTextView and coordinates EditTexts errors status.
     *  @param latitude   indicates whether coordinate is latitude or longitude coordinate
     *  @param coordinate represents value of coordinate in Double
     */
    @Override
    public void onCoordinateChanged(boolean latitude, Double coordinate) {
        // when coordinate is NaN (i.e. when empty EditText)
        if (Double.isNaN(coordinate)) {
            clearCoordinateInputOutOfRangeError(latitude);
            setTitleTextView(getString(R.string.point_north_title), Color.BLACK);
            // set null location, so needle will reset its state
            compassToLocationProvider.setTargetLocation(null);
            return;
        }

        if (latitude) {
            customTargetLocation.setLatitude(coordinate);
        } else {
            customTargetLocation.setLongitude(coordinate);
        }
        // even when coordinate is out its of range, then target location will be set to null (reset)
        compassToLocationProvider.setTargetLocation(customTargetLocation.getLocation());

        if (customTargetLocation.isCorrect()) {
            setTitleTextView(getString(R.string.point_location_title), Color.BLACK);
            clearCoordinateInputOutOfRangeError(latitude);
        }
        // correct location doesn't mean that input value is incorrect!
        if (!Utils.isCoordinateInRange(coordinate, latitude)) {
            onEmptyOrWrongInput(latitude, true);
        }
    }

    @Override
    public void setLayoutElementsOnProvider(boolean enabled) {
        setCoordinatesEditTextEnabled(enabled);
        if (enabled) {
            setSubtitleTextView(getString(R.string.info_text_subtitle), null);
            if (customTargetLocation.isCorrect()) {
                compassToLocationProvider.setTargetLocation(customTargetLocation.getLocation());
                setTitleTextView(getString(R.string.point_location_title), Color.BLACK);
            }
        } else {
            setTitleTextView(getString(R.string.point_north_title), Color.BLACK);
            setSubtitleTextView(getString(R.string.touch_info_error_subtitle), subtitleOnClickListener);
        }
    }

    private void buildAndShowLocationServicesDialog() {
        AlertDialog.Builder dialog = new AlertDialog.Builder(getActivity(), R.style.AlertDialogTheme);
        dialog.setTitle(getString(R.string.title_alert_dialog));
        dialog.setMessage(getString(R.string.message_alert_dialog));
        dialog.setPositiveButton(getString(R.string.positive_alert_dialog),
                new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface paramDialogInterface, int paramInt) {
                        Intent myIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                        startActivityForResult(myIntent, REQUEST_CODE_SETTINGS);
                    }
                });
        dialog.setNegativeButton(getString(R.string.negative_alert_dialog),
                new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface paramDialogInterface, int paramInt) {
                    }
                });

        dialog.show();
    }

    private void setCoordinatesEditTextEnabled(boolean value) {
        latitudeEditText.setEnabled(value);
        longitudeEditText.setEnabled(value);
    }

    private void setTitleTextView(String text, int color) {
        titleTextView.setText(text);
        titleTextView.setTextColor(color);
    }

    private void setSubtitleTextView(String text, View.OnClickListener onClickListener) {
        subtitleTextView.setText(text);
        subtitleTextView.setOnClickListener(onClickListener);
    }

    private void onEmptyOrWrongInput(boolean latitude, boolean outOfRange) {
        setTitleTextView(getString(R.string.point_north_title), Color.BLACK);
        if (outOfRange) {
            setCoordinateInputOutOfRangeError(latitude);
        }
    }

    private void setCoordinateInputOutOfRangeError(boolean latitude) {
        if (latitude) {
            latitudeTextInputLayout.setError(getString(R.string.error_latitude_out_of_range));
        } else {
            longitudeTextInputLayout.setError(getString(R.string.error_longitude_out_of_range));
        }
    }

    private void clearCoordinateInputOutOfRangeError(boolean latitude) {
        if (latitude) {
            latitudeTextInputLayout.setError(null);
            latitudeTextInputLayout.setErrorEnabled(false);
        } else {
            longitudeTextInputLayout.setError(null);
            longitudeTextInputLayout.setErrorEnabled(false);
        }
    }
}

File: app/src/main/kotlin/mariuszbaleczny/compass/CompassFragment.kt
package mariuszbaleczny.compass

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.provider.Settings
import android.support.annotation.StringRes
import android.support.design.widget.TextInputLayout
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.support.v4.app.LoaderManager.LoaderCallbacks
import android.support.v4.content.Loader
import android.support.v7.app.AlertDialog
import android.text.Editable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.ImageView
import android.widget.TextView
import com.pawegio.kandroid.d
import com.pawegio.kandroid.find
import com.pawegio.kandroid.textWatcher
import mariuszbaleczny.compass.custom.CustomEditText
import mariuszbaleczny.compass.location.CompassLocationPointer
import mariuszbaleczny.compass.mvp.CompassMvp
import mariuszbaleczny.compass.mvp.CompassPresenter

/**
 * Created by mariusz on 03.11.16.
 */
class CompassFragment : Fragment(), /*CompassLocationPointer.CompassToLocationListener*/ CompassMvp.View {
    companion object {
        const val COMPASS_APPLICATION: String = "compass_location_provider"
        const val REQUEST_CODE_SETTINGS: Int = 0
        const val COMPASS_LOADER_ID: Int = 100
        val TAG: String = "CompassFragment"
    }

    private var presenter: CompassMvp.Presenter? = null
    //    private var locationHelper: LocationHelper? = null
    private var compassPointer: CompassLocationPointer? = null

    private var compassView: CompassRotateHelper? = null
    private var title: TextView? = null
    private var subtitle: TextView? = null
    private var latitude: CustomEditText? = null
    private var longitude: CustomEditText? = null
    private var latitudeLayout: TextInputLayout? = null
    private var longitudeLayout: TextInputLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        loaderManager.initLoader(COMPASS_LOADER_ID, savedInstanceState, object : LoaderCallbacks<CompassMvp.Presenter> {
            override fun onLoadFinished(loader: Loader<CompassMvp.Presenter>?, data: CompassMvp.Presenter?) {
                onPresenterLoad(data!! as CompassPresenter)
            }

            override fun onCreateLoader(id: Int, args: Bundle?): Loader<CompassMvp.Presenter> {
                return CompassPresenterLoader(context)
            }

            override fun onLoaderReset(loader: Loader<CompassMvp.Presenter>?) {
                presenter = null
            }

        })

//        if (Utils.isCompassSensorPresent(context)) {
//            setupCompassToLocationProvider()
//            locationHelper = LocationHelper(Location(COMPASS_APPLICATION))
//        }
    }

    private fun onPresenterLoad(presenter: CompassPresenter) {
        this.presenter = presenter
        this.presenter?.bindView(this)
        if (compassPointer == null) compassPointer = CompassLocationPointer(context)
        this.presenter?.setCompassPointer(compassPointer!!)
        compassPointer?.startIfNotStarted()
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v: View? = inflater?.inflate(R.layout.fragment_compass, container, false)
        setupLayoutWidgets(v)
        setupTitleAndSubtitle()
        setupCompassView(v)
        setupCoordinateLayouts()
        setCoordinateInputDisabled()
//        setCoordinatesEnabled(false)
        return v
    }

    override fun asFragment(): Fragment {
        return this
    }

    override fun onResume() {
        super.onResume()
        setupLayoutOnLocationServicesCheckUp()
//        compassPointer?.startIfNotStarted()
//        checkUpAndSetupLocationServices()
    }

    override fun onPause() {
        super.onPause()
        compassPointer?.stopIfStarted()
    }

    override fun setCoordinateInputEnabled() {
        latitude?.isEnabled = true
        longitude?.isEnabled = true
    }

    override fun setCoordinateInputDisabled() {
        latitude?.isEnabled = false
        longitude?.isEnabled = false
    }

    override fun setSubtitle(@StringRes resId: Int) {
        subtitle?.setText(resId)
    }

    override fun setTitle(resId: Int) {
        title?.setText(resId)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        retainInstance = true
    }

    override fun rotateCompass(roseAngle: Int, needleAngle: Int) {
        compassView?.rotateRose(roseAngle)
        compassView?.rotateNeedle(needleAngle)
    }

    /*override fun onCompassPointerRotate(roseAngle: Int, needleAngle: Int) {
        compassView?.rotateRose(roseAngle)
        compassView?.rotateNeedle(needleAngle)
    }

    override fun setLayoutElementsOnProvider(enabled: Boolean) {
        setCoordinatesEnabled(enabled)
        if (enabled) {
            setSubtitle(R.string.info_text_subtitle, null)
            if (locationHelper?.isCorrect() as Boolean) {
                compassPointer?.setTargetLocation(locationHelper?.getLocation())
                setTitle(R.string.point_location_title, Color.BLACK)
            }
        } else {
            setTitle(R.string.needle_free_mode, Color.BLACK)
            setSubtitle(R.string.touch_info_error_subtitle, View.OnClickListener { setupLayoutOnLocationServicesCheckUp() })
        }
    }*/

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            CompassLocationPointer.LOC_PERMISSION_ON_START_REQUEST_CODE -> {
                if (Utils.areGranted(grantResults)) {
                    compassPointer?.startIfNotStarted()
                } else {
                    setCoordinateInputDisabled()
                }
            }
        }
    }

    private fun setupLayoutOnLocationServicesCheckUp() {
        if (Utils.isLocationServicesEnabled(context)) {
            setTitle(R.string.needle_free_mode)
            compassPointer?.startIfNotStarted()
        } else {
            setTitle(R.string.empty)
            setSubtitle(R.string.touch_info_error_subtitle,
                    View.OnClickListener { setupLayoutOnLocationServicesCheckUp() })
            compassPointer?.stopIfStarted()
            setCoordinateInputDisabled()
            buildAndShowLocationServicesDialog()
        }
    }

    private fun buildAndShowLocationServicesDialog() {
        val builder: AlertDialog.Builder = AlertDialog.Builder(context, R.style.AlertDialogTheme)
        builder.setTitle(R.string.title_alert_dialog)
                .setMessage(R.string.message_alert_dialog)
                .setPositiveButton(R.string.positive_alert_dialog,
                        { dialogInterface, i ->
                            startActivityForResult(
                                    Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
                                    REQUEST_CODE_SETTINGS
                            )
                        })
                .setNegativeButton(R.string.negative_alert_dialog, null)
                .show()
    }

//    private fun onLatitudeChanged(value: Double?) {
//        if (value == null) {
//            locationHelper?.setIncorrect()
//            changeUiOnInvalidLatitude()
//            resetNeedlePosition()
//            return
//        }
//
//        locationHelper?.setLatitude(value)
//
//        // even when coordinate is out its of range, then target location will be set to null (reset)
//        compassPointer?.setTargetLocation(locationHelper?.getLocation())
//
//        // incorrect location doesn't mean that input value is incorrect!
//        if (Utils.isLatitudeInRange(value)) {
//            clearLatitudeOutOfRangeError()
//        } else {
//            setTitle(R.string.needle_free_mode, Color.BLACK)
//            setLatitudeOutOfRangeError()
//        }
//    }

    override fun onLatitudeInRange() {
        latitudeLayout?.error = null
        latitudeLayout?.isErrorEnabled = false
    }

    override fun onLatitudeOutOfRange() {
        latitudeLayout?.error = getString(R.string.error_latitude_out_of_range)
    }

    override fun onLongitudeInRange() {
        longitudeLayout?.error = null
        longitudeLayout?.isErrorEnabled = false
    }

    override fun onLongitudeOutOfRange() {
        longitudeLayout?.error = getString(R.string.error_longitude_out_of_range)
    }

//    private fun onLongitudeChanged(value: Double?) {
//        if (value == null) {
//            locationHelper?.setIncorrect()
//            changeUiOnInvalidLongitude()
//            resetNeedlePosition()
//            return
//        }
//
//        locationHelper?.setLongitude(value)
//
//        // even when coordinate is out its of range, then target location will be set to null (reset)
//        compassPointer?.setTargetLocation(locationHelper?.getLocation())
//
//        // incorrect location doesn't mean that input value is incorrect!
//        if (Utils.isLongitudeInRange(value)) {
//            clearLongitudeOutOfError()
//        } else {
//            setTitle(R.string.needle_free_mode, Color.BLACK)
//            setLongitudeOutOfRangeError()
//        }
//    }

//    private fun changeUiOnInvalidLongitude() {
//        clearLongitudeOutOfError()
//        setTitle(R.string.needle_free_mode, Color.BLACK)
//    }

//    private fun clearLongitudeOutOfError() {
//        longitudeLayout?.error = null
//        longitudeLayout?.isErrorEnabled = false
//    }
//
//    private fun resetNeedlePosition() {
//        // set null location, so needle will reset its state
//        compassPointer?.setTargetLocation(null)
//    }
//
//    private fun changeUiOnInvalidLatitude() {
//        clearLatitudeOutOfRangeError()
//        setTitle(R.string.needle_free_mode, Color.BLACK)
//    }

//    private fun clearLatitudeOutOfRangeError() {
//        latitudeLayout?.error = null
//        latitudeLayout?.isErrorEnabled = false
//    }

    override fun onNullLatitude() {
        latitudeLayout?.error = null
        latitudeLayout?.isErrorEnabled = false
        setTitle(R.string.needle_free_mode, Color.BLACK)
    }

    override fun onNullLongitude() {
        latitudeLayout?.error = null
        latitudeLayout?.isErrorEnabled = false
        setTitle(R.string.needle_free_mode, Color.BLACK)
    }

    private fun setTitle(resId: Int, color: Int) {
        title?.setText(resId)
        title?.setTextColor(color)
    }

    private fun setSubtitle(resId: Int, listener: View.OnClickListener?) {
        subtitle?.setText(resId)
        subtitle?.setOnClickListener(listener)
    }

    private fun setupLayoutWidgets(v: View?) {
        title = v?.find<TextView>(R.id.fragment_compass_title_text_view)
        subtitle = v?.find<TextView>(R.id.fragment_compass_subtitle_text_view)
        latitude = v?.find<CustomEditText>(R.id.fragment_compass_latitude_edit_text)
        longitude = v?.find<CustomEditText>(R.id.fragment_compass_longitude_edit_text)
        latitudeLayout = v?.find<TextInputLayout>(R.id.fragment_compass_latitude_text_input)
        longitudeLayout = v?.find<TextInputLayout>(R.id.fragment_compass_longitude_text_input)
    }

    private fun setupCompassView(v: View?) {
        val needle: ImageView? = v?.find<ImageView>(R.id.fragment_compass_needle)
        val rose: ImageView? = v?.find<ImageView>(R.id.fragment_compass_rose)
        compassView = CompassRotateHelper(rose, needle)

        rose?.setOnClickListener { requestPermissionsIfNotGranted() }
        needle?.setOnClickListener { requestPermissionsIfNotGranted() }
    }

    private fun requestPermissionsIfNotGranted() {
        if (!Utils.arePermissionsGranted(context, CompassLocationPointer.COMPASS_PERMISSIONS)) {
            ActivityCompat.requestPermissions(context as CompassActivity,
                    CompassLocationPointer.COMPASS_PERMISSIONS,
                    CompassLocationPointer.LOC_PERMISSION_ON_START_REQUEST_CODE)
        }
    }

//    private fun setupCompassToLocationProvider() {
//        compassPointer = CompassLocationPointer(context)
//        compassPointer?.setCompassToLocationListener(this)
//    }

    private fun setupCoordinateLayouts() {
        if (Utils.isCompassSensorPresent(context)) {
            latitude?.setOnEditorActionListener({ v, id, event -> onCoordinateFieldAction(longitude, v, id) })
            longitude?.setOnEditorActionListener({ v, id, event -> onCoordinateFieldAction(latitude, v, id) })
            latitude?.textWatcher { afterTextChanged { s -> presenter?.onLatitudeChanged(parseDouble(s)) } }
            longitude?.textWatcher { afterTextChanged { s -> presenter?.onLongitudeChanged(parseDouble(s)) } }
        }
    }

    private fun onCoordinateFieldAction(complementaryField: CustomEditText?, v: TextView?, actionId: Int): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            controlFocus(complementaryField, v as View)
            return false
        }
        return true
    }

    private fun controlFocus(field: CustomEditText?, v: View) {
        if (!(field?.text?.isEmpty() as Boolean)) {
            Utils.hideKeyboard(v, context)
            v.clearFocus()
        } else {
            field?.requestFocus()
        }
    }

//    private fun onCoordinateChanged() {
//        val latitude: Double? = parseDouble(this.latitude?.text)
//        val longitude: Double? = parseDouble(this.longitude?.text)
//
//        onLatitudeChanged(latitude)
//        onLongitudeChanged(longitude)
//
//        if (locationHelper?.isCorrect() as Boolean) {
//            setTitle(R.string.point_location_title, Color.BLACK)
//            clearLatitudeOutOfRangeError()
//            clearLongitudeOutOfError()
//        }
//    }

    private fun parseDouble(s: Editable?): Double? {
        try {
            return s.toString().toDouble()
        } catch (ex: NumberFormatException) {
            d(tag, ex.message as String)
            return null
        }
    }
//
//    private fun setCoordinatesEnabled(enabled: Boolean) {
//        latitude?.isEnabled = enabled
//        longitude?.isEnabled = enabled
//    }

    private fun setupTitleAndSubtitle() {
        if (!Utils.isCompassSensorPresent(context)) {
            title?.setText(R.string.compass_not_detected_title)
            subtitle?.setText(R.string.empty)
        } else {
            title?.setText(R.string.needle_free_mode)
            subtitle?.setText(R.string.info_text_subtitle)
        }
    }

//    private fun setLatitudeOutOfRangeError() {
//        latitudeLayout?.error = getString(R.string.error_latitude_out_of_range)
//    }

//    private fun setLongitudeOutOfRangeError() {
//        longitudeLayout?.error = getString(R.string.error_longitude_out_of_range)
//    }

//    private fun checkUpAndSetupLocationServices() {
//        if (Utils.isCompassSensorPresent(activity)) {
//            setupLayoutOnLocationServicesCheckUp()
//        }
//    }

}